pallets/parachain-staking/src/delegation_requests.rs

```diff
@@ -311,6 +324,183 @@ impl<T: Config> Pallet<T> {
+		let mut state = <DelegatorState<T>>::get(&delegator).ok_or(<Error<T>>::DelegatorDNE)?;
+		let mut updated_scheduled_requests = vec![];
+
+		// pre-validate that all delegations have a Revoke request.
+		for bond in &state.delegations.0 {
+			let collator = bond.owner.clone();
+			let scheduled_requests = <DelegationScheduledRequests<T>>::get(&collator);
+			scheduled_requests
+				.iter()
+				.find(|req| {
+					req.delegator == delegator && matches!(req.action, DelegationAction::Revoke(_))
+				})
+				.ok_or(<Error<T>>::DelegatorNotLeaving)?;
+		}
+
+		// cancel all requests
+		for bond in state.delegations.0.clone() {
+			let collator = bond.owner.clone();
+			let mut scheduled_requests = <DelegationScheduledRequests<T>>::get(&collator);
+			Self::cancel_request_with_state(&delegator, &mut state, &mut scheduled_requests);
+			updated_scheduled_requests.push((collator, scheduled_requests));
+		}
+
+		updated_scheduled_requests
+			.into_iter()
+			.for_each(|(collator, scheduled_requests)| {
+				<DelegationScheduledRequests<T>>::insert(collator, scheduled_requests);
+			});
+
+		<DelegatorState<T>>::insert(delegator.clone(), state);
+		Self::deposit_event(Event::DelegatorExitCancelled { delegator });
+
+		Ok(().into())
+	}
+
+	/// Executes every [DelegationAction::Revoke] request for a delegator towards a collator.
+	/// Each delegation must have a [DelegationAction::Revoke] scheduled that must be allowed to be
+	/// executed in the current round, for this function to succeed.
+	pub(crate) fn delegator_execute_scheduled_revoke_all(
+		delegator: T::AccountId,
+		delegation_count: u32,
+	) -> DispatchResultWithPostInfo {
+		let state = <DelegatorState<T>>::get(&delegator).ok_or(<Error<T>>::DelegatorDNE)?;
+		ensure!(
+			delegation_count >= (state.delegations.0.len() as u32),
+			Error::<T>::TooLowDelegationCountToLeaveDelegators
+		);
+		let now = <Round<T>>::get().current;
+		let mut updated_scheduled_requests = vec![];
+
+		// pre-validate that all delegations have a Revoke request that can be executed now.
+		for bond in &state.delegations.0 {
+			let collator = bond.owner.clone();
+			let scheduled_requests = <DelegationScheduledRequests<T>>::get(&collator);
+			let request = scheduled_requests
+				.iter()
+				.find(|req| {
+					req.delegator == delegator && matches!(req.action, DelegationAction::Revoke(_))
+				})
+				.ok_or(<Error<T>>::DelegatorNotLeaving)?;
+
+			ensure!(
+				request.when_executable <= now,
+				<Error<T>>::DelegatorCannotLeaveYet
+			);
+		}
+
+		// we do not update the delegator state, since the it will be completely removed
+		for bond in &state.delegations.0 {
+			let collator = bond.owner.clone();
+			let mut scheduled_requests = <DelegationScheduledRequests<T>>::get(&collator);
+			let request_idx = scheduled_requests
+				.iter()
+				.position(|req| req.delegator == delegator)
+				.expect("must exist");
+			let request = &scheduled_requests[request_idx];
+			assert!(request.when_executable <= now);
```

Comment:
this assert is not needed no? we have checked it above. also finding the index of scheduled_requests might also be non-needed right? you have found the index above, cant we store the indices and use them here?